rfcs/0004-range-queries.md

```diff
@@ -0,0 +1,219 @@
+while leaving room for further optimization in the future.
+
+## Public API
+
+We introduce a new `Db::scan` API which accept a `DbRange` defining the range of  the scan.
+The result will be a `DbIterator` object which iterates the current records within the range.
+Users may either provide an explicit starting range in the call to `scan`, or they
+may `seek` using the returned iterator.
+
+```rust
+impl Db {
+    /// Scan a range of keys.
+    /// 
+    /// returns a `DbIterator`  
+    pub async fn scan(
+        &self,
+        range: &DbRange
+    ) -> Result<DbIterator, SlateDbError>;
+}
+
+struct DbRange {
+    lower: Bound<&[u8]>,
+    upper: Bound<&[u8]>,
+}
+
+impl DbIterator {
+    /// Get the next record in the scan.
+    /// 
+    /// returns Ok(None) when the scan is complete
+    /// returns Err(InvalidatedIterator) if the 
+    pub async fn next(
+        &mut self,
+    ) -> Result<Option(DbRecord), SlateDbError>;
+
+    /// Seek to a key ahead of the last key returned from the iterator or
+    /// the lower range bound if no records have yet been returned.
+    /// 
+    /// returns Ok(()) if the position is successfully advanced
+    /// returns SlateDbError::InvalidArgument if `lower_bound` is `Unbounded`
+    /// returns SlateDbError::InvalidArgument if the key is comes before the
+    ///  current iterator position.
+    /// returns Err(InvalidatedIterator) if the 
+    pub async fn seek(
+        &mut self,
+        lower_bound: Bound<&[u8]>
+    ) -> Result<(), SlateDbError>;
+}
+
+struct DbRecord {
+    key: &[u8],
+    value: &[u8],
+}
+```
+
+Note that the `DbIterator` may be invalidated internally if any of its referenced
+resources must be reclaimed. In this case, a call to `DbIterator::next` or 
+`DbIterator::scan` will return `SlateDb::InvalidatedError`. This is discussed in 
+more detail below.
+
+We will also provide `Db::scan_with_options`, which allows the user to specify the
+query isolation through a `ScanOptions` struct. Similar to `get_with_options`, the
+two isolation levels that will be supported are `Committed` and `Uncommitted`.
+The default when `scan` is called without explicit options will be `Committed`.
+
+```rust
+impl Db {
+    /// Scan a range of keys with the provided options.
+    ///
+    /// returns a `DbIterator`  
+    pub async fn scan_with_options(
+        &self,
+        range: &DbRange,
+        options: &ScanOptions,
+    ) -> Result<DbIterator, SlateDbError>;
+}
+
+pub struct ScanOptions {
+    /// The read commit level for read operations
+    pub read_level: ReadLevel,
+    /// The number of bytes to read ahead
+    pub read_ahead_size: usize,
+    /// Whether or not fetched blocks should be cached
+    pub cache_blocks: bool
+}
+```
+
+## Implementation
+
+A scan must iterate every table which may have keys within the query range.
+This includes the active WAL, memtables waiting to be flushed, and L0/SR tables 
```

Comment:
A _slight_ (dare I say, nitpicky) point of clarification: The active WAL (i.e. kv pairs that have been written to `wal` SSTs on object storage, but might or might not have yet been written to `compacted` in a L0 SST) data should always be inside the in-memory or on-disk memtables. So, I think we'd just need to check the memtables and `compacted` SSTs, not the in-memory WAL structure.

There is a brief period of time where a kv pair is written to the WAL SST but not yet put into the memtable, but I _think_ we hold a lock on db state during this time. This would be worth verifying. If we don't hold a lock, then, yes we'd have to scan WAL data structures as well if we wanted to expose all committed data to the iterator.

The other caveat is read-uncommitted. In that case, we'd check the in-memory WAL structure(s) as well.

_EDIT: Ah, I see you address this [here](https://github.com/slatedb/slatedb/pull/280/files#diff-e29e142d164766d35665bec115c1cef218231c4c93925961c1ce014d965f98e2R145-R147). I'll leave this comment for some additional context._

Reply:
Ack. I'll try to clarify this.